Fixed layout classes should still have resilient vtables

slavapestov opened the pull request:
To achieve this, re-organize the metadata so that the field offset vector comes immediately after the generic parameter area, instead of after the vtable. This means the vtable can change resiliently.

Note that for non-generic `@_fixed_layout` classes without resiliently-sized fields, field offsets are constant and will be inlined directly, but we need the more general access pattern to work if the `@_fixed_layout` class has generic layout.

Also, use method dispatch thunks to call vtable methods instead of accessing the vtable directly, as with resilient classes.

Since the metadata size is not known at compile time, we use the most general metadata instantiation strategy now.

slavapestov committed:
Evolution: Clean up superclass insertion tests a bit

- We don't want to support changing a root class of a class, so don't
  pretend that works. Some of these tests got removed recently in
  d8104e7e43fe3266247285e40b59caae27dd9110 but one still remained.

- For the tests that insert a non-root superclass in the inheritance
  hierarchy, also test calling a method of the derived class. This
  works now that we no longer hardcode vtable offsets and instead use
  dispatch thunks.

slavapestov committed:
AST: Change isResilient() to do the cheaper check first

slavapestov committed:
AST: Introduce ClassDecl::hasResilientMetadata()

slavapestov committed:
IRGen: Emit field offsets before vtable entries

We want to be able to define classes with a fixed storage layout,
but a resilient (opaque) vtable. If the class is also generic,
we still have to load field offsets from the metadata if they
are dependent.

So put the field offsets after the generic arguments and before
the vtable.

This is an ABI break for @_fixed_layout classes, which are
defined by the stdlib.

slavapestov committed:
IRGen: Refactor ClassLayoutBuilder a bit to help distinguish resilient-metadata from resilient-storage

We want @_fixed_layout classes to have non-resilient storage, but still have
resilient metadata.

slavapestov committed:
IRGen: @_fixed_layout classes still have resilient metadata

slavapestov commented:
@swift-ci Please test source compatibility

DougGregor commented:
@swift-ci Please test source compatibility

DougGregor commented:
@swift-ci Please test source compatibility

slavapestov committed:
Evolution: Add some tests for @_fixed_layout classes

@_fixed_layout classes have resilient vtables now, so we can add and
re-order methods.

Removing overrides is not legal though because they are subject to
devirtualization.

jrose-apple commented on lib/SIL/SILDeclRef.cpp:
```diff
@@ -928,6 +928,9 @@ SubclassScope SILDeclRef::getSubclassScope() const {
 
+  // FIXME: This is too narrow. Any class with resilient metadata should
+  // probably have this, at least for method overrides that don't add new
+  // vtable entries.
```
Yeah, this seems important for the correctness issue we're worried about.

jrose-apple replied:
Slava clarified privately that we can change this later as long as the fixed-layout classes we ship in Swift 5 don't break the rules that were there in Swift 5.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

rjmccall reviewed (approved):
Alright, this seems ok.

As I mentioned in chat, this implies that we consider the generic signature section of the metadata to be stable when the metadata is fixed-layout, which may make some things with resiliently-added default generic arguments harder.  But I think we can live with that.

slavapestov committed:
IRGen: Relax an assertion

The result type mismatch comes up when we're calling a covariant
override of a class method resiliently as well.
